docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+
+The name for this option will be any of the following based on the locations that it can be set:
+
+1. Explicit client configuration: `disableRequestCompression`
+
+2. System properties: `aws.disableRequestCompression`
+
+3. Environment variables: `AWS_DISABLE_REQUEST_COMPRESSION`
+
+4. AWS profile keys: `disable_request_compression`
+
+#### Minimum Request Compression Size Bytes Setting
+Service clients will be able to configure the minimum payload size before compression is active. This is because compression is actually not
+beneficial until a certain size in bytes. Anything below that number may typically result in an http body that is larger in size than if it was not compressed. That's why the default value for this configuration setting is 10,240. Otherwise, the setting should be an integer between 0 and 10,485,760. Also, worth noting is that this setting will be ignored if it is determined that the payload is a stream.
+
+The name for this option will be the following based on the location that it is set:
+
+1. Explicit client configuration: `requestMinCompressionSizeBytes`
+
+2. System properties: `aws.requestMinCompressionSizeBytes`
+
+3. Environment variables: `AWS_REQUEST_MIN_COMPRESSION_SIZE_BYTES`
+
+4. AWS profile keys: `request_min_compression_size_bytes`
+
+## Behavior
+
+The SDK will visit each algorithm in the order they appear in the trait until it finds one that's supported. Then it will attempt to compress the request using that algorithm. The application of the compression algorithm may be unsuccessful, in that case one more attempt will be made to apply
+the algorithm. If the application is unsuccessful still then it will try to find another algorithm that is both requested and supported. If so, then the process repeats until there are no algorithms left.
+
+**NOTE: I'm not so sure on the paragraph above about retrying the compression. Feel free to vote on removing it or leaving it in. I'm not even sure if it's something that will be needed**
+
+If we reach the end of the list of supported compression algorithms by the SDK and there is no match for any of the requested algorithms then the model is calling for an algorithm that is not yet supported by the SDK. In such case we will not fail with an exception. Instead there will be a log message indicating that the requested compression algorithm(s) is not
+supported and the request will be sent uncompressed.
+
+If there is success when attempting to compress the request then an http header will be either added to the request (containing the algorithm used) or the compression algorithm will be appended to the list of content encodings at the end of an existing header. The header being `Content-Encoding`.
```

Comment:
simplification: After compression occurs, the `Content-Encoding` header will contain the name of the compression algorithm used.